firewood/src/merkle/stream.rs

```diff
@@ -54,236 +48,301 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
+    let mut pos = visited_pos;
+    let mut first_loop = true;
+
+    loop {
+        match node.inner() {
+            New(NodeType::Leaf(leaf)) => {
+                let value = leaf.data.to_vec();
+                return Ok(Some((node.into_node(), value)));
+            }
+
+            Visited(NodeType::Leaf(_)) | Visited(NodeType::Extension(_)) => {
+                let Some((next_parent, next_pos)) = visited_parents.pop() else {
+                    return Ok(None);
                 };
 
-                (found_key, value)
+                node = ParentNode::Visited(next_parent);
+                pos = next_pos;
             }
-            _ => unreachable!(),
-        };
 
-        Poll::Ready(Some(Ok(return_value)))
+            New(NodeType::Extension(extension)) => {
+                let child = merkle.get_node(extension.chd())?;
+
+                pos = 0;
+                visited_parents.push((node.into_node(), pos));
+
+                node = ParentNode::New(child);
+            }
+
+            Visited(NodeType::Branch(branch)) => {
+                let compare_op = if first_loop {
+                    <u8 as PartialOrd>::ge
+                } else {
+                    <u8 as PartialOrd>::gt
+                };
+
+                let children = get_children_iter(branch)
+                    .filter(move |(_, child_pos)| compare_op(child_pos, &pos));
+
+                let next_child_success =
+                    next_child(merkle, children, visited_parents, &mut node, &mut pos)?;
+
+                if !next_child_success {
+                    return Ok(None);
+                }
+            }
+
+            New(NodeType::Branch(branch)) => {
+                if let Some(value) = branch.value.as_ref() {
+                    let value = value.to_vec();
+                    return Ok(Some((node.into_node(), value)));
+                }
+
+                let children = get_children_iter(branch);
+
+                let next_child_success =
+                    next_child(merkle, children, visited_parents, &mut node, &mut pos)?;
+
+                if !next_child_success {
+                    return Ok(None);
+                }
+            }
+        }
+
+        first_loop = false;
     }
 }
 
-/// Compute a key from a set of parents
-fn key_from_parents(parents: &[(ObjRef, u8)]) -> Vec<u8> {
-    parents[1..]
-        .chunks_exact(2)
-        .map(|parents| (parents[0].1 << 4) + parents[1].1)
-        .collect::<Vec<u8>>()
+fn get_children_iter(branch: &BranchNode) -> impl Iterator<Item = (DiskAddress, u8)> {
+    branch
+        .children
+        .into_iter()
+        .enumerate()
+        .filter_map(|(pos, child_addr)| child_addr.map(|child_addr| (child_addr, pos as u8)))
+}
+
+#[must_use]
+struct MustUse<T>(T);
+
+impl<T> From<T> for MustUse<T> {
+    fn from(t: T) -> Self {
+        Self(t)
+    }
+}
+
+impl<T: Not> Not for MustUse<T> {
+    type Output = T::Output;
+
+    fn not(self) -> Self::Output {
```

Comment:
Q: Is calling `not` sufficient in your use case for `must_use`, or does this need annotating as well?

Reply:
Yep, otherwise clippy would fail because of `-Dwarnings` in CI. I actually wasn't sure about this either, so I made sure I got the warning when I didn't "use" the `MustUse`